### ISSUE TITLE
the soft targets is wrong, do not use the soft targets

### ISSUE THREAD

**Issue by ASNVIN** [moein-shariatnia/OpenAI-CLIP#27]
> the soft targets is wrong. The model will converge to make all vectors similar, and at this point, the loss is zero.
> ```python
> soft targets：
> images_similarity = image_embeddings @ image_embeddings.T
> texts_similarity = text_embeddings @ text_embeddings.T
> targets = F.softmax(
>     (images_similarity + texts_similarity) / 2 * self.temperature, dim=-1
> )
> ```  

 _Originally posted by @ASNVIN in [#16](https://github.com/moein-shariatnia/OpenAI-CLIP/issues/16#issuecomment-2796183986)_

**Comment by khartist** [comment]
do you have any suggestion on fixing this ?

**Comment by binarydaddy** [comment]
I think an entropy regularization loss to prevent targets from becoming uniform would do the job.

**Comment by khartist** [comment]
@binarydaddy have you tried it? can you share the code and the result with me?

**Comment by ASNVIN** [comment]
@binarydaddy  Entropy regularization loss do not help. Let's review the `forward` function.
```python
    def forward(self, batch):
        # Getting Image and Text Features
        image_features = self.image_encoder(batch["image"])
        text_features = self.text_encoder(
            input_ids=batch["input_ids"], attention_mask=batch["attention_mask"]
        )
        # Getting Image and Text Embeddings (with same dimension)
        image_embeddings = self.image_projection(image_features)
        text_embeddings = self.text_projection(text_features)

        # Calculating the Loss
        logits = (text_embeddings @ image_embeddings.T) / self.temperature
        images_similarity = image_embeddings @ image_embeddings.T
        texts_similarity = text_embeddings @ text_embeddings.T
        targets = F.softmax(
            (images_similarity + texts_similarity) / 2 * self.temperature, dim=-1
        )
        texts_loss = cross_entropy(logits, targets, reduction='none')
        images_loss = cross_entropy(logits.T, targets.T, reduction='none')
        loss =  (images_loss + texts_loss) / 2.0 # shape: (batch_size)
        return loss.mean()
``` 
The reason why the model don't work is that the targets change during training. The model can easily converge and at this point, images_similarity is `torch.ones(batch_size)` matrix and texts_similarity  is `torch.ones(batch_size)` matrix, and the logits is also `torch.ones(batch_size)` matrix. The loss is 0.

@khartist  the author used such targets, because the dataset he uses has multiple captions for a single image. If you don't use such dataset, `torch.arange(batch_size)` as targets is ok.

**Comment by binarydaddy** [comment]
@ASNVIN 
I am aware that the trivial solution will result in loss of 0. We just need to guide the model away from returning same embeddings for all images / texts.

Thus, in order to prevent the model from cheating, entropy regularization will make sure the "targets" deviate from uniform distribution (which is the case when loss is 0, or images_similarity is `torch.ones(batch)`.

I think there is one more issue with current code snippet as well. The "targets" are taking in gradients from the loss function, which makes model even more prone to cheating. "detach()" on targets will help.

So combining everything, the implementation could be something like,

```python
        # Calculating the Loss
        logits = (text_embeddings @ image_embeddings.T) / self.temperature
        images_similarity = image_embeddings @ image_embeddings.T
        texts_similarity = text_embeddings @ text_embeddings.T
        target = F.softmax(
            (images_similarity + texts_similarity) / 2 * self.temperature, dim=-1
        )

        targets_detached = targets.clone().detach()

        texts_loss = cross_entropy(logits, targets_detached, reduction='none')
        images_loss = cross_entropy(logits.T, targets_detached.T, reduction='none')

        entropy_loss = torch.mean(-targets * torch.log(targets), dim=-1) # loss function to reduce entropy of targets

        loss =  (images_loss + texts_loss) / 2.0 + entropy_loss * scale # scale needed to balance out the losses
        return loss.mean()
```
FYI: I haven't tested this loss function, it is simply a theorycrafted loss.
